### PR TITLE
Add option for setting memory limit for user job containers

### PR DESCRIPTION
**Pull request by koct9i**
This adds controller agent operation options:
"set_container_memory_limit" and "container_memory_overhead".

### CODE REVIEW

File: yt/yt/server/node/exec_node/job_environment.cpp [hunk 1]
```diff
@@ -1186,6 +1186,11 @@ class TCriJobEnvironment
 
         spec->Resources.CpuLimit = config->ContainerCpuLimit;
 
+        if (config->ContainerMemoryLimit) {
+            // Sum up limits because job-proxy lives in the same container.
```
Comment (gritukan) [inline yt/yt/server/node/exec_node/job_environment.cpp:1190]: job proxy

File: yt/yt/server/node/exec_node/job.cpp [hunk 2]
```diff
@@ -2394,6 +2394,11 @@ TJobProxyInternalConfigPtr TJob::CreateConfig()
         }
     }
 
+    if (UserJobSpec_) {
+        proxyConfig->ContainerMemoryLimit = UserJobSpec_->container_memory_limit();
+        proxyConfig->JobProxyMemoryLimit = UserJobSpec_->job_proxy_memory_limit();
```
Comment (gritukan) [inline yt/yt/server/node/exec_node/job.cpp:2399]: Job proxy memory limit is useful even if there is no user job

File: yt/yt/server/lib/job_proxy/config.h [hunk 3]
```diff
@@ -249,6 +249,10 @@ class TJobProxyInternalConfig
 
     std::optional<double> ContainerCpuLimit;
 
+    i64 ContainerMemoryLimit = 0;
```
Comment (gritukan) [inline yt/yt/server/lib/job_proxy/config.h:252]: Do we need initialization here?

File: yt/yt/ytlib/controller_agent/proto/job.proto [hunk 4]
```diff
@@ -254,6 +254,8 @@ message TUserJobSpec
 
     optional double container_cpu_limit = 53 [default = 0];
 
+    optional int64 container_memory_limit = 72;
```
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/ytlib/controller_agent/proto/job.proto:257]: After second thought, this should include job proxy memory limit.
Enforcing memory limit for them together makes sense even if they are isolated.

File: yt/yt/server/node/exec_node/job.cpp (hunk 2 above)
Comment (koct9i) [reply yt/yt/server/node/exec_node/job.cpp:2399]: If we limit them both together then job proxy without user job could be limited too.

File: yt/yt/server/lib/job_proxy/config.h (hunk 3 above)
Comment (koct9i) [reply yt/yt/server/lib/job_proxy/config.h:252]: This means unlimted by default. Probably it's better wrap it with std::optional.

File: yt/yt/ytlib/controller_agent/proto/job.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/ytlib/controller_agent/proto/job.proto:257]: Maybe rename it into slot_memory_limit?

File: yt/yt/server/lib/job_proxy/config.h (hunk 3 above)
Comment (gritukan) [reply yt/yt/server/lib/job_proxy/config.h:252]: Yes

File: yt/yt/ytlib/controller_agent/proto/job.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (gritukan) [reply yt/yt/ytlib/controller_agent/proto/job.proto:257]: Seems like a good idea

**Review by k-pogorelov — CHANGES_REQUESTED** [review]

File: yt/yt/ytlib/controller_agent/proto/job.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (koct9i) [reply yt/yt/ytlib/controller_agent/proto/job.proto:257]: I'm talking about enforcing overall limit for job-proxy and user-job together.
Including all initializaion stuff and job shell.

Name "slot_contianer_memory_limit" should be fine.

**Review by k-pogorelov — APPROVED** [review]

**Comment by k-pogorelov** [comment]
Rebase this PR, please.
There is build error.

**Comment by koct9i** [comment]
@k-pogorelov done

**Pull request closed without merging**